### DISCUSSION TITLE
Default Values / Values should be available when using SSR

### DISCUSSION THREAD

**Discussion by j** [react-hook-form/react-hook-form#10603] [General]
When you useForm on the server with default values set, you get a flash of values.

Can we populate the form's values on server render to prevent this?

**Comment by j** [comment]
https://codesandbox.io/p/sandbox/vigorous-glade-vltq2f Example in astro

**Comment by leapful** [comment]
@j 

It seems you are mis-understanding how Astro hydration works and also react-hook-form works. 

You are using template directive `client:load` therefore Astro will try to perform HTML server-side rendering for your Form component, meaning it will produce HTML markup to the browser first. At that time, there is no javascript execution (react-hook-form handling) to inject the value therefore it will be flashed.

A workaround for your issue is to use template directive `client:only="react"` to render Form component and only shows when it's already rendered successfully.

Working demo: https://codesandbox.io/p/sandbox/naughty-benji-2t66hz?file=%2Fsrc%2Fpages%2Findex.astro%3A8%2C30

**Comment by j** [comment]
I fully understand how Astro hydration works.  This issue isn’t just Astro,
but next.js pages and app directory, and any server rendered then hydrated
component. It’s more obvious in app directory in both development and
production builds. Pages directory isn’t noticeable until a production
build.  React code is still executed for hydration, just doesn’t have
access to “ref”, which is how RHF is setting values.  I tested setting
defaultValue in the register function within react-hook-form and it was
able to actually server render the form as well as hydrate correctly.

Astro client:load performs server rendering with client hydration,
similarly to how next.js handles it.

This issue will be anywhere you do a react server render and use
react-hook-form, unless you explicitly pass “defaultValue” to all form
inputs.

If RHF could set defaultValue, there wouldn’t be a flash.

I want the server render to contain the form with pre-populated form then
for interactivity and benefits of RHF to take over after.

This effect becomes even worse for inputs with placeholders as you see
placeholder flash, for custom radios, ranges, etc.

Obviously using client:only or rendering form components only on the client
gets rid of it, but so many people use RHF and server render.

Even react documentation says to use defaultValue on inputs for
uncontrolled components, so I think RHF can probably set them directly if
defaultValues isn’t async.

**Comment by leapful** [comment]
@j 

`react-hook-form` depends on real DOM element to register and then manipulate form field data, that's why `ref` is being used for the initialization of form field element, including setting `defaultValue`.

If you really want to apply default value on server-side rendering, you need to do it manually. For example, using `form.control._defaultValues` to set default value when trigger `form.register`.

Working demo: https://codesandbox.io/p/sandbox/wizardly-haze-498fds?file=%2Fsrc%2Fcomponents%2FForm.tsx%3A25%2C57

**Reply by j** [reply]
That seems to fix SSR and I still don't understand why this wouldn't be the default behavior.  My team complains right now about the flashing form so we build a loader with a delay.  Our forms are highly complicated, though, but it is an annoyance I've had.  I've been using strictly SSR frameworks for the last few years and this has bugged me every time.

**Reply by simontaisne** [reply]
Hey @leapful, thanks for providing the details and workaround. I have the same question though, is there any specific reason why this couldn't be made the default behavior?

**Reply by simontaisne** [reply]
More context here https://github.com/react-hook-form/react-hook-form/issues/8707. I wonder why the RHF team isn't considering this anymore.

**Comment by alebelcor** [comment]
Might I suggest setting your default value both with `useForm({ defaultValues: {} })` and in your JSX with the `defaultValue` prop.

As far as I understand, setting it via the `defaultValue` prop will make it so after SSR the correct value is rendered. Then—after hydration—RHF will take over with its `defaultValues`.

I also experienced the flash you mention and this made it go away for me. Hopefully it helps you as well.
